src/astro/databases/base.py

```diff
@@ -0,0 +1,259 @@
+
+from astro.constants import (
+    DEFAULT_CHUNK_SIZE,
+    AppendConflictStrategy,
+    ExportExistsStrategy,
+    FileLocation,
+    FileType,
+    LoadExistStrategy,
+)
+from astro.sql.tables import Table
+from astro.utils.file import get_filetype
+from astro.utils.path import get_location, get_transport_params
+
+
+class BaseDatabase(ABC):
+    """
+    Base class to represent all the Database interactions.
+
+    The goal is to be able to support new databases by adding
+    a new module to the `astro/databases` directory, without the need of
+    changing other modules and classes.
+
+    The exception is if the Airflow connection type does not match the new
+    Database module name. In that case, we should update the dictionary
+    `CUSTOM_CONN_TYPE_TO_MODULE_PATH` available at `astro/databases/__init__.py`.
+    """
+
+    _create_schema_statement: str = "CREATE SCHEMA IF NOT EXISTS {}"
+    _drop_table_statement: str = "DROP TABLE IF EXISTS {}"
+    _create_table_statement: str = "CREATE TABLE IF NOT EXISTS {} AS {}"
+
+    def __init__(self, conn_id: str):
+        self.conn_id = conn_id
+
+    @property
+    def hook(self) -> BaseHook:
+        """
+        Return an instance of the database-specific Airflow hook.
+        """
+        raise NotImplementedError
+
+    @property
+    def connection(self) -> sqlalchemy.engine.base.Connection:
+        """
+        Return a Sqlalchemy connection object for the given database.
+        """
+        return self.sqlalchemy_engine.connect()
+
+    @property
+    def sqlalchemy_engine(self) -> sqlalchemy.engine.base.Engine:
+        """
+        Return Sqlalchemy engine.
+        """
+        return self.hook.get_sqlalchemy_engine()
+
+    def run_sql(self, sql_statement: str, parameters: Optional[dict] = None):
+        """
+        Return the results to running a SQL statement.
+
+        Whenever possible, this method should be implemented using Airflow Hooks,
+        since this will simplify the integration with Async operators.
+
+        :param sql_statement: Contains SQL query to be run against database
+        :param parameters: Optional parameters to be used to render the query
+        """
+        return self.hook.run(sql_statement, parameters)
+
+    # ---------------------------------------------------------
+    # Table metadata
+    # ---------------------------------------------------------
+    def get_table_qualified_name(self, table: Table) -> str:
+        """
+        Return table qualified name. This is Database-specific.
+        For instance, in Sqlite this is the table name. In Snowflake, however, it is the database, schema and table
+
+        :param table: The table we want to retrieve the qualified name for.
+        """
+        # Initially this method belonged to the Table class.
+        # However, in order to have an agnostic table class implementation,
+        # we are keeping all methods which vary depending on the database within the Database class.
```

Comment:
This KIND OF makes sense though it feels kind of clunky. Perhaps there's some way where the table can still own the function but takes the database as an argument? I'm not gonna die on this hill but something feels odd about the inheritance here.

Reply:
@dimberman, I know exactly how you feel about it - I had the same feeling of disgust and pain :joy:  It would be great that table owned this property. In my original implementation, that's where it belonged:
https://github.com/astro-projects/astro/pull/320/commits/b306e8df50b55a49c479095a280f6ccdfcd96b56#diff-33f4b62fcc766d55fbfd03d22eb256a156aae2cdf81977b5ec501880b6a7729eR19-R45

One way we may be able to bring this method back to the `Table` class is if we try to create the qualified name based on the metadata that was set for the `Table` - ignoring database types. I'm not sure if this will cause any issues in the future, but it may work:
a) if there is no metadata, return `name`
b) if there is `metadata.schema`, without `metadata.database`, return `schema.name`
c) if both `metadata.schema` and `metadata.database` are declared, return `database.schema.name`

Would this implicit behavior be good enough for all the currently supported databases and Redshift?

The benefit of having this in the Database class is that we stick to having database-specific customizations explicit and declared in a single class/module.